pkg/page/form.go

```diff
@@ -0,0 +1,87 @@
+package page
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/PuerkitoBio/goquery"
+	"github.com/versent/saml2aws/pkg/provider"
+)
+
+
+type Form struct {
+	URL    string
+	Method string
+	Values *url.Values
+}
+
+func (form *Form) BuildRequest() (*http.Request, error) {
+	values := strings.NewReader(form.Values.Encode())
+	req, err := http.NewRequest(form.Method, form.URL, values)
+	if err != nil {
+		return nil, errors.Wrap(err, "error building request")
+	}
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	return req, nil
+}
+
+func (form *Form) Submit(client *provider.HTTPClient) (*http.Response, error) {
+	if req, err := form.BuildRequest(); err != nil {
+		return nil, errors.Wrap(err, "error building request")
+	} else if res, err := client.Do(req); err != nil {
+		return nil, errors.Wrap(err, "error submitting form")
+	} else {
+		return res, nil
+	}
+}
+
+// If the document has multiple forms, the first form with an `action` attribute will be parsed.
+// You can specify the exact form using a CSS filter.
+func NewFormFromDocument(doc *goquery.Document, formFilter string) (*Form, error) {
+	form := Form{}
+
+	if formFilter == "" {
+		formFilter = "form[action]"
+	}
+	formSelection := doc.Find(formFilter).First()
+	if formSelection.Size() != 1 {
+		return nil, fmt.Errorf("could not find form")
+	}
+
+	if v, ok := formSelection.Attr("action"); !ok {
+		return nil, fmt.Errorf("could not extract form action")
+	} else {
+		form.URL = v
+	}
+
+	if v, ok := formSelection.Attr("method"); ok {
+		form.Method = strings.ToUpper(v)
+	} else {
+		form.Method = "POST"
+	}
+
+	form.Values = &url.Values{}
+	formSelection.Find("input").Each(func(_ int, s *goquery.Selection) {
+		if name, ok := s.Attr("name"); !ok {
```

Comment:
Again this could be trimmed down to remove the else, I would seperate out `name, ok := s.Attr("name")` then do the if statements to "flatten the code".